### DISCUSSION TITLE
A possible dumb question: is textX suitable to create a DSL for generating Django model code?

### DISCUSSION THREAD

**Discussion by simkimsia** [textX/textX#333] [Q&A]
I am thinking of writing my own DSL to generate Django model code or changes to Django model code faster.

Currently, I think like a business analyst of what the enduser requires and then I write it out in Django code (mostly models and serializers and viewsets via Django Rest Framework and Dynamic Rest). 

I typically use Django like backend and react as frontend.

I have been googling like mad hence I come across this library. I like to start with a narrow scope and see if there's a way to have a DSL that generates Django model code.

Is textX suitable and are there any examples which I can reverse engineer (to some extent) towards my objectives?

Thank you

**Comment by igordejanovic** [comment]
Yes, textX is perfectly suitable for the task. Actually, that is something it is created for. To make a clean little language that fits your needs in describing your requirements/solution and than generate a bunch of stuff for various target platforms. See the Entity example is the docs. It's very simplistic but would provide you a good starting point.

I see the success of Django and similar frameworks (like ROR for example) mainly due to use of internal DSLs (django models is a DSL that is interpreted on-the-fly to provide ORM etc., or admin is interpreted on the fly to provide CRUDS interface). With textX you can go further by having an external DSL where you can fully control your syntax and you can generate the rest of the stuff. Down the road, if you decide to leave Django, just change your generators and regenerate for the new platform.

**Comment by simkimsia** [comment]
Thanks for answering my question. I will mark this as closed then. :)

**Comment by igordejanovic** [comment]
We like to keep these "question" tagged issues open for better visibility. :)

**Comment by Aluriak** [comment]
You may want to agregate them as a FAQ in the wiki instead, to keep issues about… well, issues :)

If you agree with this, i can take care of translating current questions into a wiki in few hours.

**Comment by simkimsia** [comment]
That’s a great idea 💡

KimSia https://simkimsia.com ( https://simkimsia.com/ )

P.S. What I'm doing now: https://simkimsia.com/now

**Comment by igordejanovic** [comment]
@Aluriak Compiling some notable and often asked questions in the form of FAQ is a good idea but still we need a way for a general questions/discussion in the forum format. In the past we use google groups but GitHub provides a much nicer markdown based formatting and linking. I think that labeling issues is the best thing we can do for now as sometimes users don't know if something is an issue or they didn't understood the docs or the way textX works.

If you are willing to work on the FAQ compilation, I would really appreciate it!

**Comment by simkimsia** [comment]
Please allow me to chip in as well :)

**Comment by igordejanovic** [comment]
Great, thanks to both of you! Wiki is handled by git also but unfortunately [there is no way to accept PRs without some hacks](https://github.com/isaacs/github/issues/846) at the moment. Maybe the best approach would be to add FAQ page to the [official docs](http://www.igordejanovic.net/textX/stable/) which are also written in markdown using [MkDocs](https://www.mkdocs.org/) and are [part of this repo](https://github.com/igordejanovic/textX/tree/master/docs). Then you could send PRs as usual to this repo.

**Comment by simkimsia** [comment]
Let’s do that 🙌

KimSia kimsia@oppoin.com https://simkimsia.com ( https://simkimsia.com/ )

P.S. What I'm doing now: https://simkimsia.com/now

**Comment by igordejanovic** [comment]
Ok, let's sync your work. I suggest that one of you guys make an initial version of the page and than everyone can contribute. Who is applying for the initial version? :)

**Comment by Aluriak** [comment]
I'm applying for the initial version.

IIRC, there is some way to get some peoples to work on the wiki, without opening it for all users. This would come in handy, because the PR system sounds really heavy for just text writing.

Getting the FAQ in the docs makes them far from the issues (while wiki is litteraly 1 click away), so it maybe fit well with your will to keep questions in the current format.

When starting to write this message, i preferred the FAQ-in-wiki way, but in the end, i agree with the FAQ-in-doc.
I will make a first push soon.

**Comment by igordejanovic** [comment]
According to [this](https://help.github.com/articles/changing-access-permissions-for-wikis/) either collaborators or everyone with GitHub account can edit wiki.

It looks to me that this is the best way at the moment, it is good for someone to review the change before merging, and it is convenient that everything is in the same repo with the full history of changes. I'll add a link to the FAQ page latter from the wiki for easier access.

Thanks much for the initial version.

**Comment by simkimsia** [comment]
Thanks for doing this, @Aluriak Looks great!

I recommend having just one place for questions. Instead of having the questions in issue tracker and in stackoverflow.

Of course, @igordejanovic has the final say how he wants to do this.

If it helps with promotion, I don't mind replicating the more objective questions and answers Lucas wrote in the faq into the SO.

I have started to watch the `textx` tag in SO.

Thank you for allowing us to contribute, @igordejanovic 🙏

**Comment by igordejanovic** [comment]
BTW, SO is rarely used for questions about textX up until now. I still follow `textx` tag there and answer the questions but I'm not sure how many textX users have an SO account and would actively used it for Q&A. 
The `question` tag here in issues started organically as users started to ask questions here although there was a note in the docs for asking on SO.

SO could be a good platform for support as that's what it is made for. Maybe we could try again to direct users there for general questions? Not sure about this one.

**Comment by Aluriak** [comment]
I propose to move this conversation to a specific issue, namely #139.
